src/main/java/it/pagopa/ecommerce/commons/utils/warmup/ControllerWarmup.java

```diff
@@ -0,0 +1,76 @@
+package it.pagopa.ecommerce.commons.utils.warmup;
+
+import it.pagopa.ecommerce.commons.annotations.WarmupMethod;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.event.ContextRefreshedEvent;
+import org.springframework.stereotype.Component;
+import org.springframework.util.ClassUtils;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Controller warmup logic. This class is an {@link ApplicationListener} for the
+ * {@link ContextRefreshedEvent} raised when an ApplicationContext gets
+ * initialized or refreshed. Once the event is fired the classpath is scanned
+ * searching for all {@link RestController} an, for each of those, searching for
+ * methods annotated with {@link WarmupMethod} to be executed
+ *
+ * @see WarmupMethod
+ * @see ContextRefreshedEvent
+ * @see ApplicationListener
+ */
+@Component
+@Slf4j
+public class ControllerWarmup implements ApplicationListener<ContextRefreshedEvent> {
+
+    /**
+     * Callback method that handles {@link ContextRefreshedEvent} event
+     *
+     * @param event the event to respond to
+     */
+    @Override
+    public void onApplicationEvent(ContextRefreshedEvent event) {
+        event
+                .getApplicationContext()
+                .getBeansWithAnnotation(RestController.class)
+                .values()
+                .forEach(this::warmupController);
+    }
+
+    private void warmupController(Object controller) {
+        Class<?> controllerClass = ClassUtils.getUserClass(controller.getClass());
+        AtomicInteger warmUpMethods = new AtomicInteger();
```

Comment:
Why is an atomic needed?

Reply:
because counter is used inside a lambda expression and values used inside a lamda expression has to be final/effectively final

Reply:
The invocation logic here is sequential, so you could forgo atomics and replace `forEach` with a `mapToInt` that returns 1 and then `.sum()` (or some other combinators like [here](https://stackoverflow.com/questions/30451284/counting-elements-of-a-stream))

Reply:
However this begs the question: once we make these changes it should be trivial to make the stream parallel, so what if we made the warmup invocations parallel? Timings would be less useful though

Reply:
Make warmup method parallels and add a specific tests to check effective parallel execution. 
I leave the single warmup method elapsed time log that can be used to check whose api call is the slower etc.
See https://github.com/pagopa/pagopa-ecommerce-commons/pull/70/commits/b3d6d300dd39cd908390de1cc10fab944eccd59f